### ISSUE TITLE
NavLink should have a core set of classes and inactive ones rather than just tacking on "Active" ones.

### ISSUE THREAD

**Issue by thalaeg** [dotnet/aspnetcore#37765]
Currently the `NavLink` (https://github.com/dotnet/aspnetcore/blob/main/src/Components/Web/src/Routing/NavLink.cs) has a property called "ActiveClass". It should also have a property named "InactiveClass" so you can have a "core set" and then your inactive and active classes.

Here is an example.
Say I have a link that would look like this when inactive

```
<a class="text-gray bg-dark-gray" />
```
But when active I want it to look like 
```
<a class="text-white bg-dark-gray" />
```
In other words, I would like to change the color of the text depending if it is active or not. I am proposing the following look of component.
```
<NavLink class="bg-dark-gray" ActiveClass="text-white" InactiveClass="text-gray" />
```

This is easily acomplishable by the following changes to `NavLink`.

1. Adding a parameter
![image](https://user-images.githubusercontent.com/21165278/138360375-8b066fa0-c35e-4843-b38f-aa643ebb6f96.png)
2. changing `UpdateCssClass`
![image](https://user-images.githubusercontent.com/21165278/138360504-af60b88e-850b-4962-802a-77bd21b3374a.png)

This does **NOT** break any current functionality as people would likely just be tacking on the classes to the end. 
The issue with taking them on at the end is if you are using a CSS file, order matters within that for application.

I am happy to make a PR with this small change, that can help the NavLink be better!

Doesn't look any tests (that are currently written) are applicable to this. So only the component would need to be changed.

**Comment by kyleweishaupt** [comment]
Just saw this as I'm trying to use Tailwind CSS and Tailwind UI with Blazor. Some of the navigation styling requires different classes when inactive and this could be a very helpful feature.

**Comment by phmatray** [comment]
Should all Tailwind CSS users create their own NavLink-derived class?

There's been a lot of talk about integrating tailwind into Blazor. Most of them fail to address issues such as this one, animation handling, and the famous extra div problem. For the time being, it seems almost impossible to integrate TailwindUI components into Blazor.

**Comment by codymullins** [comment]
Looks like the original changes proposed by @thalaeg aren't quite the same now due to the code having changed. 

We've "forked" the [NavLink over to our library](https://github.com/pureblazor/components/blob/main/src/Pure.Blazor.Components/Navigation/PureNavLink.cs) and made the adjustments to display the InactiveClass, in case anyone needs it or wants to reference it for their own.

**Comment by reiv** [comment]
Specifically for Tailwind, you can also use arbitrary variants to look for the `aria-current` attribute, resulting in this monstrosity of a class:

`[&:not([aria-current='page'])]:hover:bg-primary`

**Comment by bzd3y** [comment]
For anybody banging their head against the wall over this 2 years later, the link @codymullins posted seems to now be dead, but I think this is what they linked 2 years ago before it was stripped out of their repository (maybe due to licensing?): https://github.com/pureblazor/components/blob/124441e437add12318cf7617c7815a4dfced7691/src/Pure.Blazor.Components/Navigation/PureNavLink.cs

I also came up with a quick and perhaps dirty (minimally tested) workaround that just inherits from `NavLink` and doesn't require copying code or reproducing functionality. With that being said, because of the way the class is designed, it wasn't possible to change its behavior the normal way, e.g. by overriding `UpdateCssClass()`.

```csharp
public class ActualNavLink : NavLink
{
	[Inject] private NavigationManager NavigationManager { get; set; } = default!;

	protected override void OnParametersSet()
	{
		if (string.IsNullOrWhiteSpace(InactiveClass) == false)
		{
			if (ShouldMatch(NavigationManager.Uri) == false)
			{
				const string ClassAttribute = "class";
				string? classValue = InactiveClass;

				if (AdditionalAttributes?.TryGetValue(ClassAttribute, out object? value) == true && value is not null)
				{
					classValue = $"{value} {InactiveClass}".ToString(CultureInfo.InvariantCulture);
				}

				if (AdditionalAttributes is IDictionary<string, object> attributeDictionary)
				{
					attributeDictionary[ClassAttribute] = classValue; // I'm a little wary of modifying the incoming attributes, but I'm also a little wary of NOT doing it.
				}
				else if (AdditionalAttributes is not null)
				{
					AdditionalAttributes = new Dictionary<string, object>(AdditionalAttributes)
					{
						[ClassAttribute] = classValue
					};
				}
				else
				{
					AdditionalAttributes = new Dictionary<string, object>()
					{
						[ClassAttribute] = classValue
					};
				}
			}
		}

		base.OnParametersSet();
	}

	[Parameter]
	public string? InactiveClass { get; set; }
}
```

I think I might try to do a PR to fix this. But my question for maintainers, is: is fixing the inheritance issue within the scope of that PR?

**Comment by codymullins** [comment]
@bzd3y yes that looks right (we had a rebranding) but now that repo is inactive. It did work for us for a while when we used it.

**Comment by bzd3y** [comment]
That makes sense. Your solution might have been the best. Mine works when it is on a layout, but when it is inside of something like a "menu" component it doesn't work because it relies on `OnParametersSet()` getting called, but when it is inside of a component it doesn't get called again. Oh, well.

Of course, all of this could be avoided if `NavLink` just had this property or had at least been implemented properly.

**Comment by bzd3y** [comment]
As a result of the limitation that I described above, I came up with something else. The only problem might be that it does some calculating in `BuildRenderTree()`. I don't know if that is discouraged or not.

```csharp
public class ActualNavLink : NavLink
{
	private readonly NavigationManager _navigationManager;
	private string? _cssClass;

	public ActualNavLink(NavigationManager navigationManager)
	{
		_navigationManager = navigationManager;
	}

	protected override void OnParametersSet()
	{
		if (string.IsNullOrWhiteSpace(InactiveClass) == false)
		{
			const string ClassAttribute = "class";

			if (AdditionalAttributes?.TryGetValue(ClassAttribute, out object? value) == true && value is string classValue)
			{
				_cssClass = $"{classValue} {InactiveClass}".ToString(CultureInfo.InvariantCulture);
			}
			else
			{
				_cssClass = InactiveClass;
			}
		}

		base.OnParametersSet();
	}

	protected override void BuildRenderTree(RenderTreeBuilder builder)
	{
		if (ShouldMatch(_navigationManager.Uri) == false)
		{
			CssClass = _cssClass;
		}

		base.BuildRenderTree(builder);
	}

	[Parameter]
	public string? InactiveClass { get; set; }
}
```
